### ISSUE TITLE
pyright 1.1.308+ runs indefinitely on hydra-zen source file

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#5163]
**Describe the bug**
Starting with pyright 1.1.308 (and up through 1.1.310), I am no longer able to scan the source code of [hydra-zen](https://github.com/mit-ll-responsible-ai/hydra-zen). pyright "spins" seemingly indefinitely on `src/hydra_zen/structured_configs/_make_custom_builds.py `

**To Reproduce**
```console
$ git clone git@github.com:mit-ll-responsible-ai/hydra-zen.git
$ cd hydra-zen
$ git checkout cdc8ba6f276e
$ pyright src/hydra_zen/structured_configs/_make_custom_builds.py
```

**Comment by erictraut** [comment]
Here's a minimal repro of the problem:

```python
from typing import Any, Callable, TypeVar

T = TypeVar("T", bound=Callable[..., Any])

def func1(x: T) -> T:
    return x

func1(func1)
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.

**Comment by rsokl** [comment]
Unfortunately it appears that this issue has not been resolved after upgrading to pyright 1.3.13. `pyright src/hydra_zen/structured_configs/_make_custom_builds.py` still takes a very long time to run. It looks like my [pyright CI job is crashing with an OOM](https://github.com/mit-ll-responsible-ai/hydra-zen/actions/runs/5204722994/jobs/9389318763?pr=491).

**Comment by erictraut** [comment]
The infinite recursion behavior was fixed at one point, but a subsequent change broke it again. The hydra-zen code is creating a recursive (self-referential) type when it calls `builds(builds, , **_new_defaults)`, and this results in a combinatoric explosion when performing subsequent type analysis operations. My initial fix detected and prevented the creation of this recursive type, but the detection logic was not sufficiently robust, so it broke with a subsequent change, and the break was not caught by the unit test I added with my earlier fix.

I'll explore alternate (more robust) mechanisms to detect and prevent this.

I'm also working on a mechanism by which we can run regression tests in the pyright CI on public repos that use pyright for type checking. The hydra-zen project would be a good candidate to add to this so we can avoid future regressions.
